### ISSUE TITLE
Locksmithd.service fails on startup

### ISSUE THREAD

**Issue by crawford** [coreos/bugs#838]
<a href="https://github.com/tnolet"><img src="https://avatars.githubusercontent.com/u/3802923?v=3" align="left" width="96" height="96" hspace="10"></img></a> **Issue by [tnolet](https://github.com/tnolet)**
_Saturday Jun 14, 2014 at 12:31 GMT_
_Originally opened as https://github.com/coreos/locksmith/issues/20_

----

Hi,

just updated my Vagrant box to the latest and started up a cluster of three boxes. On all boxes locksmithd.service fails to start, stating it can't connect to etcd. However, etcd starts up fine. Here is the journalctl output. 

<pre>
core@core-03 ~ $ journalctl -u locksmithd -u etcd
-- Logs begin at Sat 2014-06-14 11:38:29 UTC, end at Sat 2014-06-14 12:28:49 UTC. --
Jun 14 11:38:37 localhost systemd[1]: Starting Cluster reboot manager...
Jun 14 11:38:37 localhost systemd[1]: Started Cluster reboot manager.
Jun 14 11:38:37 localhost locksmithd[2957]: locksmithd starting currentOperation="UPDATE_STATUS_IDLE" strategy="best-effort"
Jun 14 11:38:42 localhost locksmithd[2957]: Unlocking old locks failed: etcd.service is inactive. Retrying in 10s.
Jun 14 11:38:52 core-03 locksmithd[2957]: Unlocking old locks failed: etcd.service is inactive. Retrying in 20s.
Jun 14 11:39:07 core-03 systemd[1]: Starting etcd...
Jun 14 11:39:07 core-03 systemd[1]: Started etcd.
Jun 14 11:39:07 core-03 etcd[3453]: [etcd] Jun 14 11:39:07.225 INFO      | The path /var/lib/etcd/log is in btrfs
Jun 14 11:39:07 core-03 etcd[3453]: [etcd] Jun 14 11:39:07.226 INFO      | Set NOCOW to path /var/lib/etcd/log succeeded
Jun 14 11:39:07 core-03 etcd[3453]: [etcd] Jun 14 11:39:07.227 INFO      | Discovery via https://discovery.etcd.io using prefix /65819e9544d35969a3ac50ee193a3cbb.
Jun 14 11:39:07 core-03 systemd[1]: Stopping locksmithd.service...
Jun 14 11:39:07 core-03 systemd[1]: locksmithd.service: main process exited, code=exited, status=2/INVALIDARGUMENT
Jun 14 11:39:07 core-03 systemd[1]: Stopped locksmithd.service.
Jun 14 11:39:07 core-03 systemd[1]: Unit locksmithd.service entered failed state.
Jun 14 11:39:08 core-03 etcd[3453]: [etcd] Jun 14 11:39:08.642 INFO      | Discovery found peers [http://172.17.8.102:7001]
Jun 14 11:39:08 core-03 etcd[3453]: [etcd] Jun 14 11:39:08.644 INFO      | Discovery fetched back peer list: [172.17.8.102:7001]
Jun 14 11:39:08 core-03 etcd[3453]: [etcd] Jun 14 11:39:08.649 INFO      | Send Join Request to http://172.17.8.102:7001/join
Jun 14 11:39:08 core-03 etcd[3453]: [etcd] Jun 14 11:39:08.656 INFO      | 5bdafd88c8634113b2c31a70bdf1381f joined the cluster via peer 172.17.8.102:7001
Jun 14 11:39:08 core-03 etcd[3453]: [etcd] Jun 14 11:39:08.657 INFO      | etcd server [name 5bdafd88c8634113b2c31a70bdf1381f, listen on :4001, advertised url http://172.17.8.103:4001]
Jun 14 11:39:08 core-03 etcd[3453]: [etcd] Jun 14 11:39:08.658 INFO      | peer server [name 5bdafd88c8634113b2c31a70bdf1381f, listen on :7001, advertised url http://172.17.8.103:7001]
Jun 14 11:39:08 core-03 etcd[3453]: [etcd] Jun 14 11:39:08.659 INFO      | 5bdafd88c8634113b2c31a70bdf1381f starting in peer mode
Jun 14 11:39:08 core-03 etcd[3453]: [etcd] Jun 14 11:39:08.660 INFO      | 5bdafd88c8634113b2c31a70bdf1381f: state changed from 'initialized' to 'follower'.
Jun 14 11:39:08 core-03 etcd[3453]: [etcd] Jun 14 11:39:08.709 INFO      | 5bdafd88c8634113b2c31a70bdf1381f: peer added: 'e66a184627d0481d8a89962080ccd875'
Jun 14 11:41:58 core-03 etcd[3453]: [etcd] Jun 14 11:41:58.726 INFO      | 5bdafd88c8634113b2c31a70bdf1381f: warning: heartbeat near election timeout: 2.424968433s

</pre>

**Comment by crawford** [comment]
<a href="https://github.com/philips"><img src="https://avatars.githubusercontent.com/u/33544?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [philips](https://github.com/philips)**
_Monday Jun 16, 2014 at 00:18 GMT_

----

@tnolet Does it restart later?

**Comment by crawford** [comment]
<a href="https://github.com/philips"><img src="https://avatars.githubusercontent.com/u/33544?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [philips](https://github.com/philips)**
_Monday Jun 16, 2014 at 00:20 GMT_

----

@tnolet I ask because  it should restart after 10 seconds: https://github.com/coreos/locksmith/blob/master/systemd/locksmithd.service#L15

**Comment by crawford** [comment]
<a href="https://github.com/tnolet"><img src="https://avatars.githubusercontent.com/u/3802923?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [tnolet](https://github.com/tnolet)**
_Monday Jun 16, 2014 at 09:14 GMT_

----

@philips 

no, it does not seem to restart itself. This is a dump from the status just moments ago (I left it running over the weekend):

<pre>
core@core-01 ~ $ sudo systemctl status locksmithd
● locksmithd.service
   Loaded: masked (/dev/null)
   Active: failed (Result: exit-code) since Sat 2014-06-14 12:40:44 UTC; 20h ago
 Main PID: 2978 (code=exited, status=2)

Jun 14 12:37:47 localhost locksmithd[2978]: locksmithd starting currentOperation="UPDATE_STATUS_IDLE" strategy="best-effort"
Jun 14 12:37:52 localhost locksmithd[2978]: Unlocking old locks failed: etcd.service is inactive. Retrying in 10s.
Jun 14 12:38:02 core-01 locksmithd[2978]: Unlocking old locks failed: etcd.service is inactive. Retrying in 20s.
Jun 14 12:38:22 core-01 locksmithd[2978]: Unlocking old locks failed: etcd.service is inactive. Retrying in 40s.
Jun 14 12:39:02 core-01 locksmithd[2978]: Unlocking old locks failed: etcd.service is inactive. Retrying in 1m20s.
Jun 14 12:40:22 core-01 locksmithd[2978]: Unlocking old locks failed: etcd.service is inactive. Retrying in 2m40s.
Jun 14 12:40:44 core-01 systemd[1]: Stopping locksmithd.service...
Jun 14 12:40:44 core-01 systemd[1]: locksmithd.service: main process exited, code=exited, status=2/INVALIDARGUMENT
Jun 14 12:40:44 core-01 systemd[1]: Stopped locksmithd.service.
Jun 14 12:40:44 core-01 systemd[1]: Unit locksmithd.service entered failed state.
</pre>

Then I did a `vagrant reload` and checked it again. No luck:

<pre>
core@core-01 ~ $ sudo systemctl status locksmithd
● locksmithd.service
   Loaded: masked (/dev/null)
   Active: failed (Result: exit-code) since Mon 2014-06-16 09:11:09 UTC; 2min 47s ago
 Main PID: 2971 (code=exited, status=2)

Jun 16 09:11:08 core-01 systemd[1]: Started Cluster reboot manager.
Jun 16 09:11:08 core-01 locksmithd[2971]: locksmithd starting currentOperation="UPDATE_STATUS_IDLE" strategy="off"
Jun 16 09:11:09 core-01 systemd[1]: Stopping locksmithd.service...
Jun 16 09:11:09 core-01 systemd[1]: locksmithd.service: main process exited, code=exited, status=2/INVALIDARGUMENT
Jun 16 09:11:09 core-01 systemd[1]: Stopped locksmithd.service.
Jun 16 09:11:09 core-01 systemd[1]: Unit locksmithd.service entered failed state.
</pre>

**Comment by crawford** [comment]
<a href="https://github.com/jonboulle"><img src="https://avatars.githubusercontent.com/u/2601015?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [jonboulle](https://github.com/jonboulle)**
_Monday Jun 16, 2014 at 17:00 GMT_

----

And to confirm, etcd is available? (e.g. `etcdctl ls /` succeeds)

**Comment by crawford** [comment]
<a href="https://github.com/tnolet"><img src="https://avatars.githubusercontent.com/u/3802923?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [tnolet](https://github.com/tnolet)**
_Monday Jun 16, 2014 at 21:47 GMT_

----

@jonboulle 

etcd is running just fine.

<pre>
Connection to 127.0.0.1 closed.
Tims-MacBook-Pro:vamp_templates tim$ vagrant ssh core-01
Last login: Mon Jun 16 21:43:45 2014 from 10.0.2.2
CoreOS (alpha)
Update Strategy: No Reboots
Failed Units: 1
  locksmithd.service
core@core-01 ~ $ etcdctl ls
/coreos.com
/vamp
</pre>

**Comment by crawford** [comment]
<a href="https://github.com/jonboulle"><img src="https://avatars.githubusercontent.com/u/2601015?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [jonboulle](https://github.com/jonboulle)**
_Monday Jun 16, 2014 at 21:51 GMT_

----

Doh - I missed that locksmithd is masked. Did you do that manually or through cloud-config? What's `ls -l /etc/systemd/system/locksmithd.service /run/systemd/system/locksmithd.service` give?

**Comment by crawford** [comment]
<a href="https://github.com/tnolet"><img src="https://avatars.githubusercontent.com/u/3802923?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [tnolet](https://github.com/tnolet)**
_Monday Jun 16, 2014 at 22:45 GMT_

----

Yeah, it's masked. I also can't restart it myself.

I start up vagrant boxes using a one of the Vagrantfiles from the CoreOS Vagrant repo. In it, I use cloudconfig. Please see the exact userdata file and the output for the `ls -l`. You can also see I stop the docker.service before starting it again to set the TCP socket. I can't seem to do it any other way. Using the `Before` and `After` hooks don't work. But I should log another issue for that.


<pre>
core@core-01 ~ $ ls -l /etc/systemd/system/locksmithd.service
ls: cannot access /etc/systemd/system/locksmithd.service: No such file or directory
core@core-01 ~ $ ls -l /run/systemd/system/locksmithd.service
lrwxrwxrwx 1 root root 9 Jun 16 22:37 /run/systemd/system/locksmithd.service -> /dev/null
core@core-01 ~ $ sudo systemctl restart locksmithd.service
Failed to restart locksmithd.service: Unit locksmithd.service is masked.
</pre>

userdata
<pre>
#cloud-config
coreos:
  etcd:
      #generate a new token for each unique cluster from https://discovery.etcd.io/new
      discovery: https://discovery.etcd.io/95b345ae0e4251493afeca24eb1a1842
      addr: $public_ipv4:4001
      peer-addr: $public_ipv4:7001
  update:
    reboot-strategy: off
  units:
      # this is a dirty hack. We stop docker.service before we kickstart it again using the docker-tcp.service. We need
      # this because otherwise docker.service already starts before we can assign the TCP socket
      # todo: Docker should not bind to 0.0.0.0.
    - name: docker.service
      command: stop
    - name: docker-tcp.socket
      command: start
      enable: true
      content: |
        [Unit]
        Description=Docker Socket for the API
        Before=docker.service
        [Socket]
        ListenStream=4243
        Service=docker.service
        BindIPv6Only=both
        [Install]
        WantedBy=sockets.target
    - name: etcd.service
      command: start
    - name: fleet.service
      command: start
      runtime: no
      content: |
        [Unit]
        Description=fleet
        [Service]
        Environment=FLEET_PUBLIC_IP=$public_ipv4
        ExecStart=/usr/bin/fleet
    - name: vamp_bootstrap.service
      command: start
      runtime: true
      content: |
        [Unit]
        Description=vamp_bootstrap
        After=docker.service
        After=etcd.service
        After=fleet.service
        [Service]
        TimeoutStartSec=500
        EnvironmentFile=/etc/environment
        ExecStartPre=/bin/sh -c "/usr/bin/docker history tnolet/vamp_ubuntu >/dev/null || /usr/bin/docker pull tnolet/vamp_ubuntu"
        ExecStart=/bin/sh -c "/usr/bin/docker run --rm --name vamp_agent -p 5701:5701 -p 5702:5702  -e PUBLIC_ADDRESS=$COREOS_PUBLIC_IPV4 -e DOCKER0_ADDRESS=`ifconfig docker0 | grep 'inet ' | awk '{print $2}'` -e PHYSICAL_HOSTNAME=%H tnolet/vamp_ubuntu /vamp-bootstrap-1.0/bin/cluster_boot.sh vamp-agent-0.1.0"
        ExecStop=/usr/bin/docker rm -f vamp_agent
</pre>

**Comment by crawford** [comment]
<a href="https://github.com/jonboulle"><img src="https://avatars.githubusercontent.com/u/2601015?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [jonboulle](https://github.com/jonboulle)**
_Monday Jun 16, 2014 at 23:48 GMT_

----

Thanks, that helps clear things up. 

The behaviour of locksmithd.service not running and being masked is expected, because you have `reboot-strategy: off` in the cloudconfig. However, when cloudinit is trying to stop the unit, it's reporting it as failed because locksmithd currently doesn't cleanly shutdown in response to SIGTERM if it's still trying to talk to etcd. Filed #21 to address this.

**Comment by crawford** [comment]
<a href="https://github.com/jonboulle"><img src="https://avatars.githubusercontent.com/u/2601015?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [jonboulle](https://github.com/jonboulle)**
_Monday Jun 16, 2014 at 23:50 GMT_

----

So, to be clear, this is a cosmetic issue for now and will be cleaned up once #21 lands. 

If you re-enable locksmithd (i.e. by having `reboot-strategy` set to something else) and experience further issues, please file another ticket. Thanks!
